### ISSUE TITLE
Feature request: Expand roll() to process Roll20's Rollable Tables and recursively execute their contents

### ISSUE THREAD

**Issue by phylll** [michael-buschbeck/mychs-macro-magic#76]
I've spent quite a few hours in recent weeks on concepts to implement the seemingly simple idea of automating the effects of critical hits and misses in combat, and it turns out to be quite the beast. Roll20's Rollable Tables provide an elegant way of randomly selecting one of several predefined items, but they effectively\* carry only a very limited payload: one line of text and one (optional) image. Since I don't just want to *tell* the player what happens and leave it up to them to roll the dice for how many rounds their injury will limit their movement by how much etc., I need *logic*. (\* "Effectively" refers to the observation that the API script Rollable Table Macros executes macros from the names of table items, but only once: there is no recursion, and only one line, so beyond the scope of Roll20's awkward and very limited inbuilt macro functionality, there is no reasonable way to execute the kind of logic I need through RTM.)

Now, I *think* there may be quite a simple solution that *only* requires one relatively simple additional feature from MMM, which would be for the `roll()` function to be able to process Roll20's rollable tables using the standard syntax *x*t[*table_name*] where *x* denotes the number of calls to rollable table *table_name*.

Here's the concept for my proposed use case.

In my main script, e.g. a script that processes an attack, I would like to make a random selection from a weighted table that my GM or I can manage externally:
```
!mmm script
[...]
!mmm   set foeName = "@{target|name}"
[...]
!mmm   set attackRoll = roll(combinedAttackFormula)
!mmm   if iscritical(attackRoll)
!mmm     do roll("1t[critical-hit-table]")
!mmm   end if
[...]
!mmm end script
```

Roll Table "critical-hit-table" looks something like this, with a weighted probability assigned to each macro:

Weight | Name
----- | -----
10 | #critHitWeaponLost
10 | #critHitSeveredArm
20 | #critHitConfused
... | and so on

The individual script fragments, let's take "critHitWeaponLost" as an example, do *not* start a new `script` block -- they are written as code fragments to fit the context of the original script, so as to maintain access to the complete scope of variables, used in this example to identify the enemy and process their location data:
```
!rem   // (calculate where the weapon falls based on enemy location and dice rolls for direction and distance)
!mmm   chat: You scored a critical hit! ${foeName} loses their weapon, which lands ${lostWeaponDistance} squares to the ${lost WeaponDirection} of them. 
!rem   // Ideally, we might even be able to bring a token for the weapon into existence at the right place.
```

Does this make sense?

**Comment by michael-buschbeck** [comment]
Makes complete sense, and I love the idea! :)

First off, `roll()` and inline rolls in MMM statements definitely need rollable-table support – I've really just glossed over that because I couldn't be bothered to check out what rollable tables really are. Thanks for the explanation – that's saved me some work ;)

Overloading `roll()` with extra magic is enticing, but I ultimately tend towards boringly explicit solutions. How about this?:

Firstly, `roll()` run on a rollable-table expression returns just the string of the rolled table entry.

Secondly, referencing a macro _at runtime_ that contains MMM code to execute seems useful even independently from rollable tables. We could introduce an `execute` command that takes an expression expected to evaluate to a macro reference:

| Line | Command | Comment
|-|-|-
| 1 | _!mmm_ **call** "#critHitConfused" | *(macro name expression)*
| 2 | _!mmm_ **call** #critHitConfused | *(constant macro name; syntax sugar)*
| 3 | _!mmm_ **call** roll("1t[critical-hit-table]")

**Comment by phylll** [comment]
Makes sense to me. However, I forgot something last night. My concept loses the useful ability for the GM to maintain the text and images of the roll table through the Roll20 GUI -- the pure data payloads would have to be hardcoded somewhere in my scripts if the only thing I'm getting from the `roll()` call is the item's name, and I'm (ab)using that for my "subroutine" call. 

Is there a simple way out? The complicated way out would be to create a new kind of data structure (list or table) which I would get from the `roll()` call and which would contain the image (I think it's a URL, not a blob -- I hope) and a text field. Then I would need the regexp engine to separate out the subroutine call from the descriptive text, since both would have to be stored in the text field. Hrmpf.

**Comment by michael-buschbeck** [comment]
The images are stored as URLs – [see API docs](https://help.roll20.net/hc/en-us/articles/360037772793-API-Objects#API:Objects-TableItem). (Which also explains the broken images we've been seeing in our games on occasion.)

So you'd like for the GM to be able to include customizable flavor text directly in the rollable table (though its Roll20 UI) in addition to some technical label that allows you to invoke certain behavior in your script – correct?

**Comment by phylll** [comment]
Yes, but... this mix doesn't sit perfectly well with me, either, I just don't have a better idea -- even if we completely separated the logic and the data, how would we connect the two without injecting some technical identifier into the roll table?

**Comment by michael-buschbeck** [comment]
What kind of logic specifically are you looking to implement – can you give a small survey of concrete examples? Maybe we can work something out from there.

**Comment by phylll** [comment]
Sure, here we go:
- x rounds in which character cannot attack: set counter into turn tracker and add token status marker
- roll for dexterity modified by whether character wears a helmet, and if roll fails, mark token as unconscious for 1d6 rounds (status marker and turn tracker)
- lose weapon: roll to determine which grid cell the weapon lands on relative to the character, ideally spawn image for the weapon and place it correctly
- weapon destroyed, subject to an additional check against attributes (is it a magic weapon?) and an additional roll of a die
- injured myself: rolls for damage, process damage (health/endurance points reduced, potential consequences down to death depending on character's previous health situation)
- injured someone else: roll damage, everything else is up to them

**Comment by michael-buschbeck** [comment]
That's informative. Definitely too involved to make into a domain-specific mini-language for the GM to include in the table.

So this leaves only implementing the logic in MMM explicitly, and us with the question of how to let the script select the appropriate choice of logic depending on the roll (without sullying the table with technical metadata).

I've checked out what metadata I get from rollable-table rolls ([see docs](https://help.roll20.net/hc/en-us/articles/360037256754-API-Chat#API:Chat-RollResultStructureEx.1)):
- `name` – presumably the single line of text you mentioned
- `avatar` – URL to an image
- `tableidx` – index of the table row selected by the roll
- `weight` – not directly useful, but I could reverse-engineer something like a percentile out of it ("rolled in the 95th percentile of this table")

Does any of this sound useful? The `tableidx` maybe?

**Comment by phylll** [comment]
Yep, sounds good. Should we go check if `tableidx` is consistent -- just to make sure that we don't run into trouble later if these are somehow virtual tables whose row indexes change if something else changes?

What do you think about the other challenge? Even if we use something other than `name` and `avatar` for the technical reference, you would still need to implement some kind of non-trivial data structure to return to me, and I would need syntax to work with it. Is that worth the effort? If not, we're back at hardcoding the texts and images, which is obviously doable -- I don't think our GM will want to change these things very often.

**Comment by michael-buschbeck** [comment]
I'm tempted to implement a shitty (but forward compatible) subset of some of the ideas developed in #6 and #9:

- Let `roll()` return a "struct" type.
- In scalar context, the value will behave like it used to (coerce to the rolled number or rollable-table entry "name", respectively).
- You can use `getprop(val,propname)` to extract struct member values – `name`, `avatar`, `tableidx` (exact names TBD, but you get the gist).
- Bonus syntax sugar, if I get around to it: You can say `myroll.name` to get a property.

Whether table entry indices are stable and reliable remains uncertain until we've tried it. Of course I'm getting table entires in _some_ order, but it's possible that this order isn't well-defined. There's no explicit ordering criterion in the backend data structure.

Does the Roll20 Rollable Table UI provide for a means to reorder table entries?

**Comment by phylll** [comment]
Nope. Order of entries looks consistent, though, per order of creation.

**Comment by phylll** [comment]
Good news: we're not the first to reference roll table entries by ID. The very actively maintained and well documented API script [Spawn](https://app.roll20.net/forum/permalink/9649899/) does that, too, and it seems to work consistently well. I've tested it in our TEST game, and it works intuitively: "sides" or entries of a roll table are numbered 1..n in the order they are created and shown in the roll table window.

**Comment by phylll** [comment]
> * Let `roll()` return a "struct" type.

Just to make sure: will this be able to manage *multiple* rolls of a roll table, as the Roll20 standard syntax would expect which treats roll tables like dice to call as many times as you like in one throw (`xt[name]` with *name* referencing the roll table and *x* the number of "throws")? Right now, I don't really see myself needing more than one roll table at a time, so I could easily live with reduced functionality here -- just flagging this.

**Comment by michael-buschbeck** [comment]
Requires #159

**Comment by michael-buschbeck** [comment]
For reference: This is what `msg.inlinerolls` looks like for a `[[2t[ranged-attack-critical-failure]]]` inline roll:

```JSON
[
    {
        "expression": "2t[ranged-attack-critical-failure]",
        "results": {
            "resultType": "sum",
            "rolls": [
                {
                    "dice": 2,
                    "results": [
                        {
                            "tableItem": {
                                "avatar": "https://s3.amazonaws.com/files.d20.io/images/213500373/oWPHLXXYdMeaDe8SfiwJdQ/med.png?1617624585",
                                "id": "-MWmo-yKQR6ydw5wKTDT",
                                "name": "Dein Geschoss ist zerbrochen. Bei magischen Geschossen wird mit 1W100 auf den höheren der beiden magischen Bonuswerte (*10%) gewürfelt ob diese auch zerstört ist. ",
                                "weight": 10
                            },
                            "tableidx": 40,
                            "v": 0
                        },
                        {
                            "tableItem": {
                                "avatar": "",
                                "id": "-MXWf9l9R29ur25iv8xY",
                                "name": "!mmm chat: You've rolled the BEST critical failure. Congrats!",
                                "weight": 50
                            },
                            "tableidx": 56,
                            "v": 0
                        }
                    ],
                    "sides": 101,
                    "table": "ranged-attack-critical-failure",
                    "type": "R"
                }
            ],
            "total": 0,
            "type": "V"
        },
        "signature": false
    }
]
```

**Comment by michael-buschbeck** [comment]
Here is another snippet:

```JSON
                    "results": [
                        {
                            "tableItem": {
                                "avatar": "",
                                "id": "-MXWf9l9R29ur25iv8xY",
                                "name": "!mmm chat: You've rolled the BEST critical failure. Congrats!",
                                "weight": 50
                            },
                            "tableidx": 96,
                            "v": 0
                        },
                        {
                            "tableItem": {
                                "avatar": "",
                                "id": "-MXWf9l9R29ur25iv8xY",
                                "name": "!mmm chat: You've rolled the BEST critical failure. Congrats!",
                                "weight": 50
                            },
                            "tableidx": 61,
                            "v": 0
                        },
                        {
                            "tableItem": {
                                "avatar": "",
                                "id": "-MXWf9l9R29ur25iv8xY",
                                "name": "!mmm chat: You've rolled the BEST critical failure. Congrats!",
                                "weight": 50
                            },
                            "tableidx": 95,
                            "v": 0
                        },
                        {
                            "tableItem": {
                                "avatar": "",
                                "id": "-MXWf9l9R29ur25iv8xY",
                                "name": "!mmm chat: You've rolled the BEST critical failure. Congrats!",
                                "weight": 50
                            },
                            "tableidx": 86,
                            "v": 0
                        },
                        {
                            "tableItem": {
                                "avatar": "",
                                "id": "-MXWf9l9R29ur25iv8xY",
                                "name": "!mmm chat: You've rolled the BEST critical failure. Congrats!",
                                "weight": 50
                            },
                            "tableidx": 65,
                            "v": 0
                        },
```

Notice the different `tableidx` values even though the same `tableItem` was selected. The table has a grand total of 5 entries, all of which carry significant weight (in the range from 10 to 50) though.

Apparently `tableidx` is the "roll" in the range from `1` to `sum(weights)` – so there are different `tableidx` values for each table item that has weight > 1 (specifically, one possible `tableidx` per weight point).

So... what to use in an MMM script to attach meaning to specific table items?

- `tableidx` requires the script to internally mirror the table weighting scheme. A bug there will cause the script to randomly behave differently from what the roll output told the user.
- `tableItem.id` requires the script author (or script configurator) to painstakingly copy out line-noise-style IDs and make sure they're associated with to the correct behavior.
- Some human- and script-readable identifier in `tableItem.name` would work, but it'd be visible in output sent to the player (unless channeled through script code and stripped). Plus, there's currently no MMM support for parsing strings.

**Idea:** MMM can look into the table's full definition (identified through the `table` field, [see API docs](https://wiki.roll20.net/API:Objects#Rollable_Table)) and generate a **table item index** value – the first table item would have index `0`, the second item `1`, and so on. This would be like what I *thought* `tableidx` was going to be, and the script author would only have to mirror the order of table items, not their weights as well.

Thoughts?

**Comment by phylll** [comment]
Now that MMM supports functions, why don't we simply create a feature that either allows script authors to hand over code to MMM for execution, and a roll table feature that returns `tableItem.name` etc. as a string? If that field holds the function call, we don't need some complicated way of reassociating a rolltable result with some ID, do we? If you think the risks of arbitrary dynamic code execution are too great, the rolltable feature could also have a flag that says "execute name field and return result"?

**Comment by michael-buschbeck** [comment]
With #148 fixed we don't even need a dedicated function to evaluate MMM code at runtime – just `!mmm chat:` it out:

```
!mmm script
!mmm     set dyncall = "foo(true, 123)"
!mmm     chat: !mmm do ${dyncall}
!mmm end script
```

That's asynchronous though – it'd execute the function call after the script sending it has completed running.

If you want to run additional code after the dynamic call and/or want to catch and process its return value, you could wrap the dynamic call into another function call:

```
!mmm script
!mmm     set dyncall = "foo(true, 123)"
!mmm     function RunAfterDynCall(result)
!mmm         chat: ${result}
!mmm     end function
!mmm     publish to sender: RunAfterDynCall
!mmm     chat: !mmm do RunAfterDynCall(${dyncall})
!mmm end script
```

Even with a dedicated `eval(code)` function that executes MMM code dynamically (and synchronously) something like that would need to be done to catch a function's return value – `!mmm` commands don't have return values after all, so it's not clear to me what an `eval()` function could/should return.

**Comment by michael-buschbeck** [comment]
...well, `eval()` could return the last evaluated script's variable stash in a struct-like manner (like that `script` struct in function bodies)…

**Comment by michael-buschbeck** [comment]
Aside:

> and a roll table feature that returns `tableItem.name` etc. as a string? If that field holds the function call, we don't need some complicated way of reassociating a rolltable result with some ID, do we?

I think I'm still not completely clear about the benefits provided by using rollable tables. (This is just personal interest; this doesn't affect my willingness to add support for them in a way that you consider useful.)

Does this mean that you're looking to use rollable tables more or less purely as a conveniently UI-editable configuration structure?

Am I correct in thinking that rollable tables don't provide any functionality that couldn't also be implemented by way of a single `[[1d100]]` roll and a bunch of `else if`s?

(I'm not discounting developer convenience at all – I just thought there might be more to it, e.g. making it possible for non-programmers to create and meaningfully edit them, e.g. to provide custom flavor text.)

**Comment by phylll** [comment]
Just trying to think this through after so long... On the side question: Main sticking point in my head has been to use something that's already there so that there is a single data set to be maintained *by the GM*, not by me, to be used both by scripts and manually whenever we run into some strange special case for which the scripts aren't well suited. I see your point, though -- by now we probably could have saved quite some time if I had simply built a long `if..else if...` tree and taken on maintenance myself. (Also, now there is less maintenance than in those early days when the GM would come up with new pictures for critical rolls every two weeks.)

**Comment by phylll** [comment]
So yeah ... gimme syntax to manipulate structs, I guess, so I can at least separate code and data *a little bit*, and I'll just write the code without rollable tables rather than wasting time on a rollable table implementation that will still be less than elegant.

Would this work?

```
function postCriticalCleanupOrWhatever(call)
  chat: !mmm do ${call}
  // do stuff
end function
publish to sender/game: postCriticalCleanupOrWhatever
...
set defenseCriticalSuccessEvents = [
  { weight: 10, label: "Mana from Heaven", desc: "foobar", imageURL: "...", call: "m3mgdDefenseCritSuccessLetManaFlowFromHeaven" },
  { weight: 15, label: "whatever", ... },
]
...
if iscritical(defenseRoll)
  set criticalSuccessEvent = myOwnRollTableReturnsASingleEventStruct(defenseCriticalSuccessEvents, roll("1d100"))
  chat: !mmm do postCriticalCleanupOrWhatever("${criticalSuccessEvent.call & "(" & criticalSuccessEvent.label & "," & criticalSuccessEvent.desc & "," & criticalSuccessEvent.imageURL & ")}")
end if
```

I couldn't hand over a `struct` data type to either function, right? So I would split up the parameters, and I do need to hand over data *to* the event-specific code.

What these functions would do is simply output and manipulating tokens and the turn tracker, so apart from errors I don't think we would absolutely need a way to *return* anything from the event-specific functions to the cleanup function. (And if push came to shove, I could use global variables to exchange data.)

**Comment by michael-buschbeck** [comment]
#159 is alive and well, if a bit more involved (that is to say: interesting!) because it requires the expression parser to grow backtracking abilities: When it's parsing `foo: bar` there's no saying if `foo` is a variable lookup or a literal identifier until the `:` comes along. All current syntax is unambiguous without any lookahead.

Adding rollable table support in a way that simply makes its result available is a lot more straightforward to implement – seems to me I should just do that first, if only because it's a core Roll20 feature and MMM shouldn't pretend otherwise.

**Comment by phylll** [comment]
I see the case for native rolltable support, and we might well need it somewhere down the line. Still, you should know that I've made some pretty fast progress -- thanks to your new structs etc. -- with implementing https://github.com/phylll/mychs-macro-magic/issues/3, and by now it looks as if it will be easier not to use rolltables for that at all.

**Comment by michael-buschbeck** [comment]
That's great to know!

So I'll keep this on the backburner and (maybe) rather focus on adding support for the turn tracker (#40) and various other odds and ends.
